code/modules/mob/living/simple_animal/revenant/revenant.dm

```diff
@@ -78,6 +81,72 @@
 	if(world.time <= next_move)
 		return
 	A.attack_ghost(src)
+	if(ishuman(A) && in_range(src, A))
+		src.Harvest(A)
+
+
+/mob/living/simple_animal/revenant/proc/Harvest(mob/living/carbon/human/target)
+	if(!src.castcheck(0))
+		return
+	if(src.draining)
+		src << "<span class='warning'>You are already siphoning the essence of a soul!</span>"
+		return
+	if(target in src.drained_mobs)
+		src << "<span class='warning'>[target]'s soul is dead and empty.</span>"
+		return
+	if(!target.stat)
+		src << "<span class='notice'>This being's soul is too strong to harvest.</span>"
+		if(prob(10))
+			target << "You feel as if you are being watched."
+		return
+	src.draining = 1
+	src.essence_drained = 2
+	src << "<span class='notice'>You search for the soul of [target].</span>"
+	if(do_after(src, 10, 3, 0, target)) //did they get deleted in that second?
+		if(target.ckey)
+			src << "<span class='notice'>Their soul burns with intelligence.</span>"
+			src.essence_drained += 2
+		if(target.stat != DEAD)
+			src << "<span class='notice'>Their soul blazes with life!</span>"
+			src.essence_drained += 2
+		else
+			src << "<span class='notice'>Their soul is weak and faltering.</span>"
+		if(do_after(src, 20, 6, 0, target)) //did they get deleted NOW?
+			switch(essence_drained)
+				if(1 to 2)
+					src << "<span class='info'>[target] will not yield much essence. Still, every bit counts.</span>"
+				if(3 to 4)
+					src << "<span class='info'>[target] will yield an average amount of essence.</span>"
+				if(5 to INFINITY)
+					src << "<span class='info'>Such a feast! [target] will yield much essence to you.</span>"
+			if(do_after(src, 30, 9, 0, target)) //how about now
+				if(!in_range(src, target))
```

Comment:
pretty sure ```do_after()``` already checks that the user (src) and the target (target) are in range.